designs/node-upgrades.md

```diff
@@ -0,0 +1,185 @@
+
+#### The User
+
+Taking advantage of finalizers present on all nodes, a user can decide to terminate their nodes at any time with `kubectl delete node`. If there are pods scheduled to these nodes, Karpenter will provision a replacement node if they cannot all schedule elsewhere. If a user wanted to upgrade a node, they can simply delete it. If upgrading serially, this just means multiple nodes could be scaling down at once. If upgrading in parallel, this would need to plug into the existing controls. In either case, Karpenter will rely on the user to manually intervene with any failed scheduling actions, as they have already manually intervened.
+
+#### Kubernetes Concepts
+
+PDBs, the `do-not-evict` pod annotation, and other finalizers are ways that users in Kubernetes can slow down upgrades. With restrictive enough PDBs or `do-not-evict` pods, upgrades may fail by exceeding their timeout period, where these two concepts could block a node from fully draining. In addition, if the user uses finalizers and does not clean them up, the old node may never terminate. These conditions could block an upgrade from completing independent of how many pods have been evicted. For this reason, the node could be fully upgraded but keep around old capacity. Since this may or may not block upgrades, we will incorporate this logic into the following section on prioritizations.
+
+## Which nodes should Karpenter upgrade?
+
+When provisioning a node, Karpenter generates a launch template and fills it with a random AMI selected by the AWSNodeTemplate AMISelector. As hinted above, nodes with AMIs that have drifted from the AWSNodeTemplate will be upgraded.
+
+A cluster may undergo less churn if Karpenter prioritizes upgrading nodes that are close to being recycled. In addition, Karpenter will also not upgrade nodes that it will be unable to upgrade successfully. With this in mind, Karpenter will not upgrade nodes that fit these conditions:
+
+* Has a pod with a do-not-evict: true annotation. Karpenter will not attempt to drain a node with a pod with this annotation. For the same reason, this will hold true for upgrades. This will be included in Upgrade Best Practices.
+* Nodes with pods that have misconfigured PDBs, or blocking PDBs altogether. If a pod is owned by two PDBs, the [eviction API blocks eviction](https://kubernetes.io/docs/concepts/scheduling-eviction/api-eviction/#how-api-initiated-eviction-works)of that pod. This requires user intervention.
+
+On top of this, Karpenter’s current AMI selection picks a random AMI out of a list of allowed AMIs. Karpenter will now implement AMI selection to get the latest AMI out of the allowed list. If users need a method of not having the latest EKS Optimized AMI, Karpenter can implement a way to include native AMI Versioning controls in the future.
```

Comment:
Side question - it picks AMIs by random from a list? There is no ordering at all, not even on AMI date?

Reply:
Currently it's at random. We'll adjust this to pick the latest, native AMI versioning in the future could include pinning versions of the EKS Optimized AMI or some sort of mandated bake time. Would be great to get feedback here as well.